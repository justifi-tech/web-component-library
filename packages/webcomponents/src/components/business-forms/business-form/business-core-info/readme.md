# justifi-business-generic-info-form



<!-- Auto Generated Below -->


## Properties

| Property         | Attribute | Description | Type             | Default     |
| ---------------- | --------- | ----------- | ---------------- | ----------- |
| `formController` | --        |             | `FormController` | `undefined` |


## Dependencies

### Used by

 - [justifi-business-form](..)

### Depends on

- [form-control-text](../../../../ui-components/form)
- [form-control-select](../../../../ui-components/form)
- [form-control-date](../../../../ui-components/form)
- [form-control-number-masked](../../../../ui-components/form)

### Graph
```mermaid
graph TD;
  justifi-business-core-info --> form-control-text
  justifi-business-core-info --> form-control-select
  justifi-business-core-info --> form-control-date
  justifi-business-core-info --> form-control-number-masked
  form-control-text --> form-control-tooltip
  form-control-tooltip --> custom-popper
  form-control-select --> form-control-tooltip
  form-control-date --> form-control-tooltip
  form-control-number-masked --> form-control-tooltip
  justifi-business-form --> justifi-business-core-info
  style justifi-business-core-info fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
