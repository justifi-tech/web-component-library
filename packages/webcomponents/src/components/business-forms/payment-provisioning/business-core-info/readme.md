# justifi-business-generic-info-form



<!-- Auto Generated Below -->


## Properties

| Property              | Attribute               | Description | Type       | Default     |
| --------------------- | ----------------------- | ----------- | ---------- | ----------- |
| `allowOptionalFields` | `allow-optional-fields` |             | `boolean`  | `undefined` |
| `authToken`           | `auth-token`            |             | `string`   | `undefined` |
| `businessId`          | `business-id`           |             | `string`   | `undefined` |
| `getBusiness`         | --                      |             | `Function` | `undefined` |
| `patchBusiness`       | --                      |             | `Function` | `undefined` |


## Events

| Event         | Description | Type                                   |
| ------------- | ----------- | -------------------------------------- |
| `error-event` |             | `CustomEvent<ComponentError>`          |
| `formLoading` |             | `CustomEvent<boolean>`                 |
| `submitted`   |             | `CustomEvent<BusinessFormSubmitEvent>` |


## Methods

### `validateAndSubmit({ onSuccess }: { onSuccess: any; }) => Promise<void>`



#### Parameters

| Name  | Type                  | Description |
| ----- | --------------------- | ----------- |
| `__0` | `{ onSuccess: any; }` |             |

#### Returns

Type: `Promise<void>`




## Dependencies

### Used by

 - [justifi-business-core-info-form-step](.)

### Depends on

- [form-control-text](../../../form)
- [form-control-select](../../../form)
- [form-control-date](../../../form)
- [form-control-number-masked](../../../form)

### Graph
```mermaid
graph TD;
  justifi-business-core-info-form-step-core --> form-control-text
  justifi-business-core-info-form-step-core --> form-control-select
  justifi-business-core-info-form-step-core --> form-control-date
  justifi-business-core-info-form-step-core --> form-control-number-masked
  form-control-text --> form-control-help-text
  form-control-text --> form-control-error-text
  form-control-select --> form-control-help-text
  form-control-select --> form-control-error-text
  form-control-date --> form-control-help-text
  form-control-date --> form-control-error-text
  form-control-number-masked --> form-control-help-text
  form-control-number-masked --> form-control-error-text
  justifi-business-core-info-form-step --> justifi-business-core-info-form-step-core
  style justifi-business-core-info-form-step-core fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
