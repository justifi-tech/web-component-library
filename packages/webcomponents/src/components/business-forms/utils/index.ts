export * from './business-form-options';
export * from './helpers';
export * from './input-handlers';
export * from './event-types';
export * from './country-form-helpers';
