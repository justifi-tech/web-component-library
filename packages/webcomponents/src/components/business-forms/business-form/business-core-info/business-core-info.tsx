import { Component, Host, h, Prop, State } from '@stencil/core';

import {
  BusinessStructureOptions,
  BusinessTypeOptions,
} from '../business-form-schema';
import { FormController } from '../../../form/form';
import { PHONE_MASKS, TAX_ID_MASKS } from '../../../../utils/form-input-masks';
import { CoreBusinessInfo, ICoreBusinessInfo } from '../../../../api/Business';

/**
 *
 * The difference between this component and business-core-info-details
 * is that this component is meant to be a form and send data
 * and the other one  is meant to be just read only.
 *
 * @exportedPart label: Label for inputs
 * @exportedPart input: The input fields
 * @exportedPart input-invalid: Invalid state for inputfs
 */
@Component({
  tag: 'justifi-business-core-info',
  styleUrl: 'business-core-info.scss',
})
export class BusinessCoreInfo {
  @Prop() formController: FormController;
  @State() errors: any = {};
  @State() coreInfo: ICoreBusinessInfo = {};

  constructor() {
    this.inputHandler = this.inputHandler.bind(this);
  }

  componentDidLoad() {
    this.formController.values.subscribe(
      values => (this.coreInfo = { ...new CoreBusinessInfo(values) }),
    );
    this.formController.errors.subscribe(errors => {
      this.errors = { ...errors };
    });
  }

  inputHandler(name: string, value: string) {
    this.formController.setValues({
      ...this.formController.values.getValue(),
      [name]: value,
    });
  }

  render() {
    const coreInfoDefaultValue = this.formController.getInitialValues();

    return (
      <Host exportparts="label,input,input-invalid">
        <fieldset>
          <legend>General Info</legend>
          <hr />
          <div class="row gy-3">
            <div class="col-12 col-md-6">
              <form-control-text
                name="legal_name"
                label="Legal Name"
                defaultValue={coreInfoDefaultValue.legal_name}
                error={this.errors.legal_name}
                inputHandler={this.inputHandler}
              />
            </div>
            <div class="col-12 col-md-6">
              <form-control-text
                name="doing_business_as"
                label="Doing Business As (DBA)"
                defaultValue={coreInfoDefaultValue.doing_business_as}
                error={this.errors.doing_business_as}
                inputHandler={this.inputHandler}
              />
            </div>
            <div class="col-12 col-md-6">
              <form-control-select
                name="business_type"
                label="Business Type"
                options={BusinessTypeOptions}
                defaultValue={coreInfoDefaultValue.business_type}
                error={this.errors.business_type}
                inputHandler={this.inputHandler}
              />
            </div>
            <div class="col-12 col-md-6">
              <form-control-select
                name="business_structure"
                label="Business Structure"
                options={BusinessStructureOptions}
                defaultValue={coreInfoDefaultValue.business_structure}
                error={this.errors.business_structure}
                inputHandler={this.inputHandler}
              />
            </div>
            <div class="col-12">
              <form-control-text
                name="industry"
                label="Industry"
                defaultValue={coreInfoDefaultValue.industry}
                error={this.errors.business_structure}
                inputHandler={this.inputHandler}
              />
            </div>
            <div class="col-12">
              <form-control-number-masked
                name="tax_id"
                label="Tax ID"
                defaultValue={coreInfoDefaultValue.tax_id}
                error={this.errors.tax_id}
                inputHandler={this.inputHandler}
                mask={TAX_ID_MASKS.US}
              />
            </div>
            <div class="col-12">
              <form-control-text
                name="website_url"
                label="Website URL"
                defaultValue={coreInfoDefaultValue.website_url}
                error={this.errors.website_url}
                inputHandler={this.inputHandler}
              />
            </div>
            <div class="col-12 col-md-6">
              <form-control-text
                name="email"
                label="Email Address"
                defaultValue={coreInfoDefaultValue.email}
                error={this.errors.email}
                inputHandler={this.inputHandler}
              />
            </div>
            <div class="col-12 col-md-6">
              <form-control-number-masked
                name="phone"
                label="Phone Number"
                defaultValue={coreInfoDefaultValue.phone}
                error={this.errors.phone}
                inputHandler={this.inputHandler}
                mask={PHONE_MASKS.US}
              />
            </div>
          </div>
        </fieldset>
      </Host>
    );
  }
}
