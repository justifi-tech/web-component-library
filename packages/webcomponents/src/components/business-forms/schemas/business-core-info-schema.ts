import { object, string } from 'yup';
import { 
  doingBusinessAsValidation,
  emailValidation,
  industryValidation,
  businessNameValidation, 
  phoneValidation, 
  taxIdValidation, 
  websiteUrlValidation,
  dateOfIncorporationValidation,
  businessClassificationValidation,
  makeBusinessClassificationValidation,
  makeTaxIdValidation
} from './schema-validations';
import { CountryCode } from '../../../utils/country-codes';
import { countryLabels } from '../utils/country-config';

// Base schema (no country-specific rules for classification/tax id)
export const baseBusinessCoreInfoSchema = (allowOptionalFields?: boolean) => {
  const schema = object({
    legal_name: businessNameValidation.required('Enter legal name'),
    website_url: websiteUrlValidation.required('Enter business website url'),
    email: emailValidation.required('Enter business email'),
    phone: phoneValidation.required('Enter phone number'),
    doing_business_as: doingBusinessAsValidation.nullable(),
    classification: businessClassificationValidation.required('Select business classification'),
    industry: industryValidation.required('Enter a business industry'),
    tax_id: taxIdValidation.required('Enter valid tax ID (SSN or EIN) without dashes'),
    date_of_incorporation: dateOfIncorporationValidation.required('Enter date of incorporation'),
  });

  const easySchema = object({
    legal_name: businessNameValidation.required('Enter legal name'),
    website_url: websiteUrlValidation.nullable(),
    email: emailValidation.nullable(),
    phone: phoneValidation.nullable(),
    doing_business_as: doingBusinessAsValidation.nullable(),
    classification: businessClassificationValidation.nullable(),
    industry: industryValidation.nullable(),
    tax_id: taxIdValidation.nullable(),
    date_of_incorporation: dateOfIncorporationValidation.nullable(),
  });

  return allowOptionalFields ? easySchema : schema;
};

// Country-specific schema convenience wrappers
export const businessCoreInfoSchemaUSA = (allowOptionalFields?: boolean) =>
  baseBusinessCoreInfoSchema(allowOptionalFields).concat(object({
    classification: makeBusinessClassificationValidation(CountryCode.USA).required('Select business classification'),
    tax_id: makeTaxIdValidation(CountryCode.USA).required('Enter valid tax ID (SSN or EIN) without dashes'),
  } as any));

export const businessCoreInfoSchemaCAN = (allowOptionalFields?: boolean) =>
  baseBusinessCoreInfoSchema(allowOptionalFields).concat(object({
    classification: makeBusinessClassificationValidation(CountryCode.CAN).required('Select business classification'),
    tax_id: makeTaxIdValidation(CountryCode.CAN).required(`Enter valid ${countryLabels.CAN.taxIdLabel} without dashes`),
  } as any));

// Back-compat default: USA
export const businessCoreInfoSchema = (allowOptionalFields?: boolean) => businessCoreInfoSchemaUSA(allowOptionalFields);
