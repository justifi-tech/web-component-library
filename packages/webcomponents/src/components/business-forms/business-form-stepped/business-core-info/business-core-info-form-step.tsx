import { Component, Host, h, Prop, State, Method, Event, EventEmitter } from '@stencil/core';
import { FormController } from '../../../form/form';
import { PHONE_MASKS, TAX_ID_MASKS } from '../../../../utils/form-input-masks';
import { BusinessFormServerErrors, CoreBusinessInfo, IBusiness, ICoreBusinessInfo } from '../../../../api/Business';
import { Api, IApiResponse } from '../../../../api';
import { config } from '../../../../../config';
import { parseCoreInfo } from '../../utils/parsers';
import { flattenNestedObject } from '../../../../utils/utils';
import { BusinessStructureOptions, BusinessTypeOptions, coreInfoSchema } from '../../schemas/business-core-info-form-schema';

/**
 *
 * The difference between this component and business-generic-info-details
 * is that this component is meant to be a form and send data
 * and the other one  is meant to be just read only.
 *
 * @exportedPart label: Label for inputs
 * @exportedPart input: The input fields
 * @exportedPart input-invalid: Invalid state for inputfs
 */
@Component({
  tag: 'justifi-business-core-info-form-step',
  styleUrl: 'business-core-info-form-step.scss',
})
export class BusinessCoreInfoFormStep {
  @Prop() authToken: string;
  @Prop() businessId: string;
  @State() formController: FormController;
  @State() errors: any = {};
  @State() coreInfo: ICoreBusinessInfo = {};
  @Event({ bubbles: true }) submitted: EventEmitter<{ data?: any}>;
  @Event() formLoading: EventEmitter<boolean>;
  @Event() serverError: EventEmitter<{ data: any, message: BusinessFormServerErrors }>;

  constructor() {
    this.inputHandler = this.inputHandler.bind(this);
    this.sendData = this.sendData.bind(this);
    this.fetchData = this.fetchData.bind(this);
  }

  private api: any;

  get businessEndpoint() {
    return `entities/business/${this.businessId}`
  }

  private async fetchData() {
    this.formLoading.emit(true);
    try {
      const response: IApiResponse<IBusiness> = await this.api.get(this.businessEndpoint);
      this.coreInfo = new CoreBusinessInfo(response.data);
      this.formController.setInitialValues({ ...this.coreInfo });
    } catch (error) {
      this.serverError.emit({ data: error, message: BusinessFormServerErrors.fetchData });
    } finally {
      this.formLoading.emit(false);
    }
  }

  private async sendData(onSuccess?: () => void) {
    this.formLoading.emit(true);
    try {
      const payload = parseCoreInfo(flattenNestedObject(this.formController.values.getValue()));
      const response = await this.api.patch(this.businessEndpoint, JSON.stringify(payload));
      this.handleResponse(response, onSuccess);
    } catch (error) {
      this.serverError.emit({ data: error, message: BusinessFormServerErrors.patchData });
    } finally {
      this.formLoading.emit(false);
    }
  }
  
  handleResponse(response, onSuccess) {
    if (response.error) {
      this.serverError.emit({ data: response.error, message: BusinessFormServerErrors.patchData });
    } else {
      onSuccess();
    }
    this.submitted.emit({ data: response });
  }

  @Method()
  async validateAndSubmit(onSuccess: () => Promise<void>) {
    this.formController.validateAndSubmit(() => this.sendData(onSuccess));
  };

  componentWillLoad() {
    const missingAuthTokenMessage = 'Warning: Missing auth-token. The form will not be functional without it.';
    const missingBusinessIdMessage = 'Warning: Missing business-id. The form requires an existing business-id to function.';
    if (!this.authToken) console.error(missingAuthTokenMessage);
    if (!this.businessId) console.error(missingBusinessIdMessage);

    this.formController = new FormController(coreInfoSchema);
    this.api = Api(this.authToken, config.proxyApiOrigin);
    this.fetchData();
  }

  componentDidLoad() {
    this.formController.values.subscribe(values => 
      this.coreInfo = { ...values }
    );
    this.formController.errors.subscribe(errors => {
      this.errors = { ...errors };
    });
  }

  inputHandler(name: string, value: string) {
    this.formController.setValues({
      ...this.formController.values.getValue(),
      [name]: value,
    });
  }

  render() {
    const coreInfoDefaultValue = this.formController.getInitialValues();

    return (
      <Host exportparts="label,input,input-invalid">
        <form>
          <fieldset>
            <legend>General Info</legend>
            <hr />
            <div class="row gy-3">
              <div class="col-12 col-md-6">
                <form-control-text
                  name="legal_name"
                  label="Legal Name"
                  defaultValue={coreInfoDefaultValue.legal_name}
                  error={this.errors.legal_name}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class="col-12 col-md-6">
                <form-control-text
                  name="doing_business_as"
                  label="Doing Business As (DBA)"
                  defaultValue={coreInfoDefaultValue.doing_business_as}
                  error={this.errors.doing_business_as}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class="col-12 col-md-6">
                <form-control-select
                  name="business_type"
                  label="Business Type"
                  options={BusinessTypeOptions}
                  defaultValue={coreInfoDefaultValue.business_type}
                  error={this.errors.business_type}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class="col-12 col-md-6">
                <form-control-select
                  name="business_structure"
                  label="Business Structure"
                  options={BusinessStructureOptions}
                  defaultValue={coreInfoDefaultValue.business_structure}
                  error={this.errors.business_structure}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class="col-12">
                <form-control-text
                  name="industry"
                  label="Industry"
                  defaultValue={coreInfoDefaultValue.industry}
                  error={this.errors.industry}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class="col-12">
                <form-control-number-masked
                  name="tax_id"
                  label="Tax ID"
                  defaultValue={coreInfoDefaultValue.tax_id}
                  error={this.errors.tax_id}
                  inputHandler={this.inputHandler}
                  mask={TAX_ID_MASKS.US}
                />
              </div>
              <div class="col-12">
                <form-control-text
                  name="website_url"
                  label="Website URL"
                  defaultValue={coreInfoDefaultValue.website_url}
                  error={this.errors.website_url}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class="col-12 col-md-6">
                <form-control-text
                  name="email"
                  label="Email Address"
                  defaultValue={coreInfoDefaultValue.email}
                  error={this.errors.email}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class="col-12 col-md-6">
                <form-control-number-masked
                  name="phone"
                  label="Phone Number"
                  defaultValue={coreInfoDefaultValue.phone}
                  error={this.errors.phone}
                  inputHandler={this.inputHandler}
                  mask={PHONE_MASKS.US}
                />
              </div>
            </div>
          </fieldset>
        </form>
      </Host>
    );
  }
}
