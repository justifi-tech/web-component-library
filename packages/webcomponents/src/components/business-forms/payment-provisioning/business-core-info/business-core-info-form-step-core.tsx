import { Component, Host, h, Prop, State, Method, Event, EventEmitter } from '@stencil/core';
import { businessCoreInfoSchema } from '../../schemas/business-core-info-schema';
import { FormController } from '../../../form/form';
import { CoreBusinessInfo, ICoreBusinessInfo } from '../../../../api/Business';
import { BusinessFormSubmitEvent } from '../../utils/business-form-types';
import { ComponentError } from '../../../../api/ComponentError';
import { businessClassificationOptions } from '../../utils/business-form-options';
import { PHONE_MASKS } from '../../../../utils/form-input-masks';
import { parseCoreInfo } from '../../utils/payload-parsers';
import { flattenNestedObject } from '../../../../utils/utils';
import { numberOnlyHandler } from '../../../form/utils';

@Component({
  tag: 'justifi-business-core-info-form-step-core',
})
export class BusinessCoreInfoFormStepCore {
  @State() formController: FormController;
  @State() errors: any = {};
  @State() coreInfo: ICoreBusinessInfo = {};

  @Prop() authToken: string;
  @Prop() businessId: string;
  @Prop() getBusiness: Function;
  @Prop() patchBusiness: Function;
  @Prop() allowOptionalFields?: boolean;

  @Event({ bubbles: true }) submitted: EventEmitter<BusinessFormSubmitEvent>;
  @Event({ bubbles: true }) formLoading: EventEmitter<boolean>;
  @Event({ eventName: 'error-event', bubbles: true }) errorEvent: EventEmitter<ComponentError>;

  @Method()
  async validateAndSubmit({ onSuccess }) {
    this.formController.validateAndSubmit(() => this.sendData(onSuccess));
  };

  get patchPayload() {
    let formValues = parseCoreInfo(flattenNestedObject(this.formController.values.getValue()));
    return JSON.stringify(formValues);
  }

  componentWillLoad() {
    this.getBusiness && this.getData();
    this.formController = new FormController(businessCoreInfoSchema(this.allowOptionalFields));
  }

  componentDidLoad() {
    this.formController.errors.subscribe(errors => {
      this.errors = { ...errors };
    });
  }

  private getData = () => {
    this.formLoading.emit(true);
    this.getBusiness({
      onSuccess: (response) => {
        this.coreInfo = new CoreBusinessInfo(response.data);
        this.formController.setInitialValues({ ...this.coreInfo });
      },
      onError: ({ error, code, severity }) => {
        this.errorEvent.emit({
          message: error,
          errorCode: code,
          severity: severity
        });
      },
      final: () => this.formLoading.emit(false)
    });
  }

  private sendData = (onSuccess: () => void) => {
    let submittedData;
    this.formLoading.emit(true);
    this.patchBusiness({
      payload: this.patchPayload,
      onSuccess: (response) => {
        submittedData = response;
        onSuccess();
      },
      onError: ({ error, code, severity }) => {
        submittedData = error;
        this.errorEvent.emit({
          message: error,
          errorCode: code,
          severity: severity
        });
      },
      final: () => {
        this.submitted.emit({ data: { submittedData } });
        this.formLoading.emit(false)
      }
    });
  }

  inputHandler = (name: string, value: string) => {
    this.formController.setValues({
      ...this.formController.values.getValue(),
      [name]: value,
    });
  }

  render() {
    const coreInfoDefaultValue = this.formController.getInitialValues();

    return (
      <Host exportparts='label,input,input-invalid'>
        <form>
          <fieldset>
            <legend>General Info</legend>
            <hr />
            <div class='row gy-3'>
              <div class='col-12'>
                <form-control-text
                  name='legal_name'
                  label='Legal Name'
                  defaultValue={coreInfoDefaultValue.legal_name}
                  errorText={this.errors.legal_name}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class='col-12'>
                <form-control-text
                  name='doing_business_as'
                  label='Doing Business As (DBA)'
                  defaultValue={coreInfoDefaultValue.doing_business_as}
                  errorText={this.errors.doing_business_as}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class='col-12 col-md-8'>
                <form-control-select
                  name='classification'
                  label='Business Classification'
                  options={businessClassificationOptions}
                  defaultValue={coreInfoDefaultValue.classification}
                  errorText={this.errors.classification}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class='col-12 col-md-4'>
                <form-control-date
                  name='date_of_incorporation'
                  label='Date of Incorporation'
                  defaultValue={coreInfoDefaultValue.date_of_incorporation}
                  errorText={this.errors.date_of_incorporation}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class='col-12 col-md-6'>
                <form-control-text
                  name='industry'
                  label='Industry'
                  defaultValue={coreInfoDefaultValue.industry}
                  errorText={this.errors.industry}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class='col-12 col-md-6'>
                <form-control-text
                  name='tax_id'
                  label='Tax ID (EIN or SSN)'
                  defaultValue={coreInfoDefaultValue.tax_id}
                  errorText={this.errors.tax_id}
                  inputHandler={this.inputHandler}
                  keyDownHandler={numberOnlyHandler}
                  maxLength={9}
                />
              </div>
              <div class='col-12'>
                <form-control-text
                  name='website_url'
                  label='Website URL'
                  defaultValue={coreInfoDefaultValue.website_url}
                  errorText={this.errors.website_url}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class='col-12 col-md-6'>
                <form-control-text
                  name='email'
                  label='Email Address'
                  defaultValue={coreInfoDefaultValue.email}
                  errorText={this.errors.email}
                  inputHandler={this.inputHandler}
                />
              </div>
              <div class='col-12 col-md-6'>
                <form-control-number-masked
                  name='phone'
                  label='Phone Number'
                  defaultValue={coreInfoDefaultValue.phone}
                  errorText={this.errors.phone}
                  inputHandler={this.inputHandler}
                  mask={PHONE_MASKS.US}
                />
              </div>
            </div>
          </fieldset>
        </form>
      </Host>
    );
  }
}
