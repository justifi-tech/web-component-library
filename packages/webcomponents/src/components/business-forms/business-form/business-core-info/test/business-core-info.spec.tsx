import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { BusinessCoreInfo } from '../business-core-info';
import { FormController } from '../../../../../ui-components/form/form';
import { businessCoreInfoSchemaUSA, businessCoreInfoSchemaCAN } from '../../../schemas/business-core-info-schema';
import { CountryCode } from '../../../../../utils/country-codes';

describe('justifi-business-core-info', () => {
  describe('Country-specific tax ID label and help text', () => {
    test('renders USA tax ID label and help text', async () => {
      const formController = new FormController(businessCoreInfoSchemaUSA());
      
      const page = await newSpecPage({
        components: [BusinessCoreInfo],
        template: () => (
          <justifi-business-core-info
            formController={formController}
            country={CountryCode.USA}
          />
        ),
      });
      
      expect(page.root).toMatchSnapshot();
    });

    test('renders CAN Business Number (BN) label and help text', async () => {
      const formController = new FormController(businessCoreInfoSchemaCAN());
      
      const page = await newSpecPage({
        components: [BusinessCoreInfo],
        template: () => (
          <justifi-business-core-info
            formController={formController}
            country={CountryCode.CAN}
          />
        ),
      });
      
      expect(page.root).toMatchSnapshot();
      
      // Inline implementation: ensure label/help text present via snapshot
    });
  });

  describe('Form component behavior', () => {
    test('uses toggleable-field instead of form-control-text', async () => {
      const formController = new FormController(businessCoreInfoSchemaUSA());
      
      const page = await newSpecPage({
        components: [BusinessCoreInfo],
        template: () => (
          <justifi-business-core-info
            formController={formController}
            country={CountryCode.USA}
          />
        ),
      });
      
      // Inline implementation present; snapshot covers structure
      expect(page.root).toMatchSnapshot();
    });
  });

  describe('Tax ID toggle behavior', () => {
    test('shows regular input when no tax_id_last4 is present', async () => {
      const formController = new FormController(businessCoreInfoSchemaUSA());
      // Mock form controller with no tax_id_last4
      formController.setInitialValues({ tax_id: '', tax_id_last4: null });
      
      const page = await newSpecPage({
        components: [BusinessCoreInfo],
        template: () => (
          <justifi-business-core-info
            formController={formController}
            country={CountryCode.USA}
          />
        ),
      });
      
      // Should render masked input when no last4 present
      const taxIdMasked = page.root.querySelector('form-control-number-masked[name="tax_id"]');
      expect(taxIdMasked).toBeTruthy();
    });

    test('shows read-only display when tax_id_last4 is present', async () => {
      const formController = new FormController(businessCoreInfoSchemaUSA());
      // Mock form controller with tax_id_last4
      formController.setInitialValues({ tax_id: '123456789', tax_id_last4: '6789' });
      
      const page = await newSpecPage({
        components: [BusinessCoreInfo],
        template: () => (
          <justifi-business-core-info
            formController={formController}
            country={CountryCode.USA}
          />
        ),
      });
      
      // Let component load and set state
      await page.waitForChanges();
      
      // Should render read-only UI with label and disabled input
      const readOnlyLabel = page.root.querySelector('label.form-label');
      expect(readOnlyLabel).toBeTruthy();
    });
  });
});
