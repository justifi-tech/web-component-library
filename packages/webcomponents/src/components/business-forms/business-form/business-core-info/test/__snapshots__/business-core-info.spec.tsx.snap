// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`justifi-business-core-info Country-specific tax ID label and help text renders CAN Business Number (BN) label and help text 1`] = `
<justifi-business-core-info>
  <fieldset>
    <legend part="heading-2 heading text color font-family">
      General Info
    </legend>
    <div class="gy-3 row">
      <div class="col-12">
        <form-control-text label="Legal Name" name="legal_name"></form-control-text>
      </div>
      <div class="col-12">
        <form-control-text label="Doing Business As (DBA)" name="doing_business_as"></form-control-text>
      </div>
      <div class="col-12 col-md-8">
        <form-control-select label="Business Classification" name="classification"></form-control-select>
      </div>
      <div class="col-12 col-md-4">
        <form-control-date label="Date of Incorporation" name="date_of_incorporation"></form-control-date>
      </div>
      <div class="col-12 col-md-6">
        <form-control-text label="Industry" name="industry"></form-control-text>
      </div>
      <div class="col-12 col-md-6">
        <form-control-number-masked helptext="Business Numbers (BN) are nine digits. Enter value without spaces or dashes." label="Business Number (BN)" mask="000-000-000" name="tax_id"></form-control-number-masked>
      </div>
      <div class="col-12">
        <form-control-text label="Website URL" name="website_url"></form-control-text>
      </div>
      <div class="col-12 col-md-6">
        <form-control-text label="Email Address" name="email"></form-control-text>
      </div>
      <div class="col-12 col-md-6">
        <form-control-number-masked label="Phone Number" mask="(000) 000-0000" name="phone"></form-control-number-masked>
      </div>
    </div>
  </fieldset>
</justifi-business-core-info>
`;

exports[`justifi-business-core-info Country-specific tax ID label and help text renders USA tax ID label and help text 1`] = `
<justifi-business-core-info>
  <fieldset>
    <legend part="heading-2 heading text color font-family">
      General Info
    </legend>
    <div class="gy-3 row">
      <div class="col-12">
        <form-control-text label="Legal Name" name="legal_name"></form-control-text>
      </div>
      <div class="col-12">
        <form-control-text label="Doing Business As (DBA)" name="doing_business_as"></form-control-text>
      </div>
      <div class="col-12 col-md-8">
        <form-control-select label="Business Classification" name="classification"></form-control-select>
      </div>
      <div class="col-12 col-md-4">
        <form-control-date label="Date of Incorporation" name="date_of_incorporation"></form-control-date>
      </div>
      <div class="col-12 col-md-6">
        <form-control-text label="Industry" name="industry"></form-control-text>
      </div>
      <div class="col-12 col-md-6">
        <form-control-number-masked helptext="Employer Identification Numbers (EINs) are nine digits. The federal tax identification number/EIN issued to you by the IRS. It can be found on your tax returns. Enter value without dashes." label="Tax ID (EIN or SSN)" mask="000-00-0000" name="tax_id"></form-control-number-masked>
      </div>
      <div class="col-12">
        <form-control-text label="Website URL" name="website_url"></form-control-text>
      </div>
      <div class="col-12 col-md-6">
        <form-control-text label="Email Address" name="email"></form-control-text>
      </div>
      <div class="col-12 col-md-6">
        <form-control-number-masked label="Phone Number" mask="(000) 000-0000" name="phone"></form-control-number-masked>
      </div>
    </div>
  </fieldset>
</justifi-business-core-info>
`;

exports[`justifi-business-core-info Form component behavior uses toggleable-field instead of form-control-text 1`] = `
<justifi-business-core-info>
  <fieldset>
    <legend part="heading-2 heading text color font-family">
      General Info
    </legend>
    <div class="gy-3 row">
      <div class="col-12">
        <form-control-text label="Legal Name" name="legal_name"></form-control-text>
      </div>
      <div class="col-12">
        <form-control-text label="Doing Business As (DBA)" name="doing_business_as"></form-control-text>
      </div>
      <div class="col-12 col-md-8">
        <form-control-select label="Business Classification" name="classification"></form-control-select>
      </div>
      <div class="col-12 col-md-4">
        <form-control-date label="Date of Incorporation" name="date_of_incorporation"></form-control-date>
      </div>
      <div class="col-12 col-md-6">
        <form-control-text label="Industry" name="industry"></form-control-text>
      </div>
      <div class="col-12 col-md-6">
        <form-control-number-masked helptext="Employer Identification Numbers (EINs) are nine digits. The federal tax identification number/EIN issued to you by the IRS. It can be found on your tax returns. Enter value without dashes." label="Tax ID (EIN or SSN)" mask="000-00-0000" name="tax_id"></form-control-number-masked>
      </div>
      <div class="col-12">
        <form-control-text label="Website URL" name="website_url"></form-control-text>
      </div>
      <div class="col-12 col-md-6">
        <form-control-text label="Email Address" name="email"></form-control-text>
      </div>
      <div class="col-12 col-md-6">
        <form-control-number-masked label="Phone Number" mask="(000) 000-0000" name="phone"></form-control-number-masked>
      </div>
    </div>
  </fieldset>
</justifi-business-core-info>
`;
